### PR TITLE
IGNITE-22124 Java thin: Implement Compute MapReduce API

### PR DESCRIPTION
**Pull request by vadimkolodingg**
https://issues.apache.org/jira/browse/IGNITE-22124

### CODE REVIEW

File: modules/client/src/main/java/org/apache/ignite/internal/client/compute/ClientJobExecution.java [hunk 1]
```diff
@@ -120,7 +120,7 @@ public CompletableFuture<R> resultAsync() {
         );
     }
 
-    private CompletableFuture<@Nullable Boolean> changePriority(UUID jobId, int newPriority) {
+    static CompletableFuture<@Nullable Boolean> changePriority(ReliableChannel ch, UUID jobId, int newPriority) {
```
Comment (PakhomovAlexander) [inline modules/client/src/main/java/org/apache/ignite/internal/client/compute/ClientJobExecution.java:123]: What is the reason for this change?

File: modules/client/src/main/java/org/apache/ignite/internal/client/compute/ClientTaskExecution.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+        jobIdFuture = reqFuture.thenApply(SubmitTaskResult::jobId);
+        jobIdsFuture = reqFuture.thenApply(SubmitTaskResult::jobIds);
+
+        resultAsync = reqFuture
+                .thenCompose(SubmitTaskResult::notificationFuture)
+                .thenApply(r -> {
+                    // Notifications require explicit input close.
+                    try (r) {
+                        R result = (R) r.in().unpackObjectFromBinaryTuple();
```
Comment (PakhomovAlexander) [inline modules/client/src/main/java/org/apache/ignite/internal/client/compute/ClientTaskExecution.java:70]: `r` is confusing name here.

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/client/ItThinClientComputeTest.java [hunk 3]
```diff
@@ -668,6 +677,42 @@ void testBigDecimalPropagation(String number, int scale) {
         assertEquals(expected, res);
     }
 
+    @Test
+    void testExecuteMapReduce() throws Exception {
+        TaskExecution<String> execution = client().compute().submitMapReduce(List.of(), MapReduceNodeNameTask.class.getName());
+
+        assertThat(execution.resultAsync(), willBe(allOf(containsString("itcct_n_3344"), containsString("itcct_n_3345"))));
```
Comment (PakhomovAlexander) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/client/ItThinClientComputeTest.java:684]: `itcct_n_3344` should not be used directly. Please, get the node name from the client and then compare.

**Review by PakhomovAlexander — COMMENTED** [review]
I'll leave this for @Pochatkin , no comments from me.

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/compute/ClientComputeExecuteMapReduceRequest.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+import org.apache.ignite.internal.client.proto.ClientMessageUnpacker;
+import org.apache.ignite.internal.compute.IgniteComputeInternal;
+import org.jetbrains.annotations.Nullable;
+
+
+/**
+ * Compute MapReduce request.
+ */
+public class ClientComputeExecuteMapReduceRequest {
+
```
Comment (ptupitsyn) [inline modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/compute/ClientComputeExecuteMapReduceRequest.java:42]:
```suggestion
```

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/compute/ClientComputeExecuteMapReduceRequest.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+import org.apache.ignite.client.handler.NotificationSender;
+import org.apache.ignite.compute.DeploymentUnit;
+import org.apache.ignite.compute.JobStatus;
+import org.apache.ignite.compute.TaskExecution;
+import org.apache.ignite.internal.client.proto.ClientMessagePacker;
+import org.apache.ignite.internal.client.proto.ClientMessageUnpacker;
+import org.apache.ignite.internal.compute.IgniteComputeInternal;
+import org.jetbrains.annotations.Nullable;
+
+
```
Comment (ptupitsyn) [inline modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/compute/ClientComputeExecuteMapReduceRequest.java:37]:
```suggestion
```

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/compute/ClientComputeExecuteMapReduceRequest.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+     * @param compute Compute.
+     * @param notificationSender Notification sender.
+     * @return Future.
+     */
+    public static CompletableFuture<Void> process(
+            ClientMessageUnpacker in,
+            ClientMessagePacker out,
+            IgniteComputeInternal compute,
+            NotificationSender notificationSender) {
+
```
Comment (ptupitsyn) [inline modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/compute/ClientComputeExecuteMapReduceRequest.java:57]:
```suggestion
```

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/compute/ClientComputeExecuteMapReduceRequest.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        TaskExecution<Object> execution = compute.submitMapReduce(deploymentUnits, taskClassName, args);
+        sendTaskResult(execution, notificationSender);
+
+        var idsAsync = execution.idsAsync()
+                .handleAsync((ids, ex) -> {
+                    // empty ids in case of split exception to properly respond with task id and failed status
+                    return ex == null ? ids : Collections.<UUID>emptyList();
+                });
+        return execution.idAsync()
```
Comment (ptupitsyn) [inline modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/compute/ClientComputeExecuteMapReduceRequest.java:70]:
```suggestion

        return execution.idAsync()
```

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/compute/ClientComputeExecuteMapReduceRequest.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+                .thenAcceptBoth(idsAsync, (id, ids) -> {
+                    out.packUuidNullable(id);
```
Comment (ptupitsyn) [inline modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/compute/ClientComputeExecuteMapReduceRequest.java:72]: We can't return a null `id` to the client from the initial execution request, it does not make sense. The id should not be null here.

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/compute/ClientComputeExecuteMapReduceRequest.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+                    packJobIds(out, ids);
+                });
+    }
+
+    static void packJobIds(ClientMessagePacker out, List<@Nullable UUID> ids) {
+        out.packInt(ids.size());
+        for (var uuid : ids) {
+            out.packUuidNullable(uuid);
```
Comment (ptupitsyn) [inline modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/compute/ClientComputeExecuteMapReduceRequest.java:80]: Same here, the job id can't be null. The client would not be able to do anything with it.

I think the API contract is wrong too:
* It is ok to have null in `statusAsync` when retention time limit is exceeded
* It is NOT ok to have a null id. `JobExecution` instance without an id does not make sense.

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/compute/ClientComputeExecuteMapReduceRequest.java [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+        }
+    }
+
+    static CompletableFuture<Object> sendTaskResult(TaskExecution<Object> execution, NotificationSender notificationSender) {
+        return execution.resultAsync().whenComplete((val, err) ->
+                execution.statusAsync().whenComplete((status, errStatus) ->
+                        execution.statusesAsync().whenComplete((statuses, errStatuses) ->
+                                notificationSender.sendNotification(w -> {
```
Comment (ptupitsyn) [inline modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/compute/ClientComputeExecuteMapReduceRequest.java:88]: `err`, `errStatus`, `errStatuses` are ignored?

File: modules/client/src/main/java/org/apache/ignite/internal/client/ClientUtils.java [hunk 11]
```diff
@@ -213,6 +213,7 @@ public static ClientOperationType opCodeToClientOperationType(int opCode) {
 
             case ClientOp.COMPUTE_EXECUTE:
             case ClientOp.COMPUTE_EXECUTE_COLOCATED:
+            case ClientOp.COMPUTE_EXECUTE_MAPREDUCE:
```
Comment (ptupitsyn) [inline modules/client/src/main/java/org/apache/ignite/internal/client/ClientUtils.java:216]: Let's have separate codes for `COLOCATED` and `MAPREDUCE`.

File: modules/client/src/main/java/org/apache/ignite/internal/client/compute/ClientCompute.java [hunk 12]
```diff
@@ -442,6 +474,31 @@ private static SubmitResult unpackSubmitResult(PayloadInputChannel ch) {
...
+     * @param ch Payload channel.
+     * @return Result of the task submission.
+     */
+    private static SubmitTaskResult unpackSubmitTaskResult(PayloadInputChannel ch) {
+        var jobId = ch.in().unpackUuid();
+
+        List<UUID> jobIds;
+        var size = ch.in().unpackInt();
+        if (size == 0) {
+            jobIds = Collections.emptyList();
```
Comment (ptupitsyn) [inline modules/client/src/main/java/org/apache/ignite/internal/client/compute/ClientCompute.java:491]: Is this possible and why?

File: modules/client/src/main/java/org/apache/ignite/internal/client/compute/SubmitTaskResult.java [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+import java.util.concurrent.CompletableFuture;
+import org.apache.ignite.internal.client.PayloadInputChannel;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Result of the task submission.
+ * Contains unpacked job id, a collection of ids of the jobs which are executing under this task, and notification future.
+ */
+class SubmitTaskResult extends SubmitResult {
+
```
Comment (ptupitsyn) [inline modules/client/src/main/java/org/apache/ignite/internal/client/compute/SubmitTaskResult.java:31]:
```suggestion
```

**Review by ptupitsyn — CHANGES_REQUESTED** [review]

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/compute/ClientComputeExecuteMapReduceRequest.java [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+            NotificationSender notificationSender) {
+        List<DeploymentUnit> deploymentUnits = unpackDeploymentUnits(in);
+        String taskClassName = in.unpackString();
+        Object[] args = unpackArgs(in);
+
+        TaskExecution<Object> execution = compute.submitMapReduce(deploymentUnits, taskClassName, args);
+        sendTaskResult(execution, notificationSender);
+
+        var idsAsync = execution.idsAsync()
+                .handleAsync((ids, ex) -> {
```
Comment (ptupitsyn) [inline modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/compute/ClientComputeExecuteMapReduceRequest.java:63]:
```suggestion
                .handle((ids, ex) -> {
```

File: modules/client/src/main/java/org/apache/ignite/internal/client/compute/ClientCompute.java [hunk 15]
```diff
@@ -442,6 +473,26 @@ private static SubmitResult unpackSubmitResult(PayloadInputChannel ch) {
...
+     * @param ch Payload channel.
+     * @return Result of the task submission.
+     */
+    private static SubmitTaskResult unpackSubmitTaskResult(PayloadInputChannel ch) {
+        var jobId = ch.in().unpackUuid();
+
+        var size = ch.in().unpackInt();
+        List<UUID> jobIds = new ArrayList<>(size);
+        for (int i = 0; i < size; i++) {
+            jobIds.add(ch.in().unpackUuidNullable());
```
Comment (ptupitsyn) [inline modules/client/src/main/java/org/apache/ignite/internal/client/compute/ClientCompute.java:490]:
```suggestion
            jobIds.add(ch.in().unpackUuid());
```

I think the server does not return nulls anymore.

File: modules/client/src/main/java/org/apache/ignite/internal/client/compute/SubmitTaskResult.java [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+import java.util.concurrent.CompletableFuture;
+import org.apache.ignite.internal.client.PayloadInputChannel;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Result of the task submission.
+ * Contains unpacked job id, a collection of ids of the jobs which are executing under this task, and notification future.
+ */
+class SubmitTaskResult extends SubmitResult {
+    private final List<@Nullable UUID> jobIds;
```
Comment (ptupitsyn) [inline modules/client/src/main/java/org/apache/ignite/internal/client/compute/SubmitTaskResult.java:31]:
```suggestion
    private final List<UUID> jobIds;
```

**Review by ptupitsyn — CHANGES_REQUESTED** [review]

**Review by ptupitsyn — APPROVED** [review]
